osmosis-labs/osmosis
# Protorev smarter logic and more testing

## What is the purpose of the change

This PR adds more testing to `x/protorev` and improves the arb calculation logic to be more efficient.


## Brief Changelog
 
  - Given a route, we now calculate if the minAmountIn is profitable as a pre-check. If minAmountIn is not profitable, then the route is not profitable and we don't run the full binary search algo. This will make Protorev take much less compute time in production compared to its previous design (which would run the full binary search even when a route has no-arb). We only increment the pool point counter if the trade is profitable, thereby saving us compute for swaps that create no backrunning opportunity.
  - Given a route, we now calculate if the maxAmountIn and maxAmountIn + 1 profit is increasing as a pre-check and increase the range if so. If the profit from maxInputIn is still increasing, this means that we need to increase the range to capture the max profit.
  - Adds tests that verifies the limit/point system works properly, >3 pool routes work, and other denom besides osmo/atom work.
  - Deprecate hot routes from being included in genesis state for simplicity. 
  - Add E2E testing that 
    - Checks that the module is correctly init on start
    - Checks that the module can correctly spot a back running opportunity after a swap
    - Checks that the module executes a profitable cyclic arbitrage trade
    - Checks that the module did not mint or alter the supplies from the bank
 - Add support for gamm swap msgs
 - Deprecate atom from the list of base denoms that are used on genesis for route creation


## Testing and Verifying

- All previous tests pass with new changes
- Added a test to ensure range extension works properly
- Added a test for swapExactAmountOut
- Added benchmark testing
- Added doomsday testing to ensure point system works properly and doesn't allow protorev to keep running
- Adds tests for generalized denoms
- Adds a test to ensure panic catching is handled properly. This currently FAILS. This is on purpose, this PR is not to be merged until https://github.com/osmosis-labs/osmosis/issues/4180 is resolved
- Add test for gamm msg swaps
- Add tests for E2E (as described above)
